File: mipaper/site_contract.py
from __future__ import annotations


def validate_daily_report_payload(report: dict) -> None:
    _require_branch_report_payload(report, required_keys=("report_date", "category", "total_papers"))
    _require_list(report, "focus_topics")
    _require_list(report, "topic_distribution")
    _require_list(report, "papers")


def validate_hf_report_payload(report: dict) -> None:
    _require_branch_report_payload(report, required_keys=("report_date", "total_papers"))
    _require_list(report, "focus_topics")
    _require_list(report, "topic_distribution")
    _require_list(report, "top_submitters")
    _require_list(report, "papers")


def validate_conference_report_payload(report: dict) -> None:
    _require_branch_report_payload(report, required_keys=("venue", "venue_series", "venue_year", "total_papers"))
    _require_list(report, "subject_distribution")
    _require_list(report, "topic_distribution")
    _require_list(report, "papers")
    _require_bool(report, "is_complete")


def validate_trending_report_payload(report: dict) -> None:
    _require_object("trending report payload", report)
    _require_string(report, "snapshot_date")
    _require_string(report, "generated_at")
    _require_string(report, "source_url")
    _require_number(report, "total_repositories")
    _require_list(report, "language_distribution")
    _require_list(report, "top_repositories")
    _require_list(report, "repositories")


def validate_daily_manifest(manifest: dict) -> None:
    _validate_branch_manifest(
        manifest,
        branch_key="daily",
        branch_label="Cool Daily",
        required_report_keys=(
            "slug",
            "report_date",
            "category",
            "total_papers",
            "classifier",
            "generated_at",
            "source_url",
            "focus_topics",
            "top_topics",
        ),
    )
    _require_list(manifest, "category_order")
    _require_list(manifest, "latest_by_category")


def validate_hf_manifest(manifest: dict) -> None:
    _validate_branch_manifest(
        manifest,
        branch_key="hf-daily",
        branch_label="HF Daily",
        required_report_keys=(
            "slug",
            "report_date",
            "total_papers",
            "classifier",
            "generated_at",
            "source_url",
            "focus_topics",
            "top_topics",
            "top_submitters",
        ),
    )


def validate_conference_manifest(manifest: dict) -> None:
    _validate_branch_manifest(
        manifest,
        branch_key="conference",
        branch_label="Conference",
        required_report_keys=(
            "slug",
            "venue",
            "venue_series",
            "venue_year",
            "total_papers",
            "classifier",
            "generated_at",
            "source_url",
            "subject_distribution",
            "top_topics",
        ),
    )


def validate_trending_manifest(manifest: dict) -> None:
    _validate_branch_manifest(
        manifest,
        branch_key="trending",
        branch_label="Trending",
        required_report_keys=(
            "slug",
            "snapshot_date",
            "since",
            "total_repositories",
            "generated_at",
            "source_url",
            "top_languages",
            "top_repositories",
        ),
    )


def validate_branch_catalog_manifest(manifest: dict) -> None:
    _require_object("branch catalog manifest", manifest)
    _require_string(manifest, "generated_at")
    _require_number(manifest, "reports_count")
    branches = _require_list(manifest, "branches")
    reports = _require_list(manifest, "reports")
    if manifest["reports_count"] != len(reports):
        raise ValueError(f"branch catalog reports_count mismatch: expected {len(reports)}, got {manifest['reports_count']}")

    for branch in branches:
        _require_object("branch catalog branch", branch)
        _require_string(branch, "branch_key")
        _require_string(branch, "branch_label")
        _require_string(branch, "default_report_path", allow_empty=True)
        _require_string(branch, "manifest_path")
        _require_number(branch, "reports_count")

    for report in reports:
        _require_object("branch catalog report", report)
        _require_string(report, "branch_key")
        _require_string(report, "branch_label")
        _require_string(report, "data_path")
        _require_string(report, "search_text")


def _validate_branch_manifest(manifest: dict, *, branch_key: str, branch_label: str, required_report_keys: tuple[str, ...]) -> None:
    _require_object(f"{branch_label} manifest", manifest)
    _require_string(manifest, "branch_key", expected=branch_key)
    _require_string(manifest, "branch_label", expected=branch_label)
    _require_string(manifest, "generated_at")
    _require_number(manifest, "reports_count")
    _require_string(manifest, "default_report_path", allow_empty=True)
    reports = _require_list(manifest, "reports")
    if manifest["reports_count"] != len(reports):
        raise ValueError(f"{branch_label} manifest reports_count mismatch: expected {len(reports)}, got {manifest['reports_count']}")

    for report in reports:
        _require_object(f"{branch_label} manifest report", report)
        _require_string(report, "branch_key", expected=branch_key)
        _require_string(report, "branch_label", expected=branch_label)
        _require_string(report, "data_path")
        for key in required_report_keys:
            if key in {"total_papers", "total_repositories"}:
                _require_number(report, key)
            elif key in {"focus_topics", "top_topics", "top_submitters", "subject_distribution", "top_languages", "top_repositories"}:
                _require_list(report, key)
            else:
                _require_string(report, key)


def _require_branch_report_payload(report: dict, *, required_keys: tuple[str, ...]) -> None:
    _require_object("branch report payload", report)
    _require_string(report, "generated_at")
    _require_string(report, "source_url")
    _require_string(report, "classifier")
    for key in required_keys:
        if key == "total_papers":
            _require_number(report, key)
        else:
            _require_string(report, key)


def _require_object(label: str, value: object) -> dict:
    if not isinstance(value, dict):
        raise ValueError(f"{label} must be an object")
    return value


def _require_list(payload: dict, key: str) -> list:
    value = payload.get(key)
    if not isinstance(value, list):
        raise ValueError(f"{key} must be a list")
    return value


def _require_string(payload: dict, key: str, *, allow_empty: bool = False, expected: str | None = None) -> str:
    value = payload.get(key)
    if not isinstance(value, str):
        raise ValueError(f"{key} must be a string")
    if not allow_empty and not value.strip():
        raise ValueError(f"{key} must not be empty")
    if expected is not None and value != expected:
        raise ValueError(f"{key} must be {expected!r}, got {value!r}")
    return value


def _require_number(payload: dict, key: str) -> int | float:
    value = payload.get(key)
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError(f"{key} must be a number")
    return value


def _require_bool(payload: dict, key: str) -> bool:
    value = payload.get(key)
    if not isinstance(value, bool):
        raise ValueError(f"{key} must be a boolean")
    return value

File: site/site_contract.js
function requireObject(label, value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${label} must be an object`);
  }
  return value;
}

function requireString(payload, key, options = {}) {
  const { allowEmpty = false, expected } = options;
  const value = payload?.[key];
  if (typeof value !== "string") {
    throw new Error(`${key} must be a string`);
  }
  if (!allowEmpty && !value.trim()) {
    throw new Error(`${key} must not be empty`);
  }
  if (expected !== undefined && value !== expected) {
    throw new Error(`${key} must be ${JSON.stringify(expected)}`);
  }
  return value;
}

function requireNumber(payload, key) {
  const value = payload?.[key];
  if (typeof value !== "number" || Number.isNaN(value)) {
    throw new Error(`${key} must be a number`);
  }
  return value;
}

function requireBoolean(payload, key) {
  const value = payload?.[key];
  if (typeof value !== "boolean") {
    throw new Error(`${key} must be a boolean`);
  }
  return value;
}

function requireArray(payload, key) {
  const value = payload?.[key];
  if (!Array.isArray(value)) {
    throw new Error(`${key} must be an array`);
  }
  return value;
}

function validateBranchManifest(manifest, options) {
  const { branchKey, branchLabel, requiredReportKeys } = options;
  requireObject(`${branchLabel} manifest`, manifest);
  requireString(manifest, "branch_key", { expected: branchKey });
  requireString(manifest, "branch_label", { expected: branchLabel });
  requireString(manifest, "generated_at");
  requireNumber(manifest, "reports_count");
  requireString(manifest, "default_report_path", { allowEmpty: true });
  const reports = requireArray(manifest, "reports");
  if (manifest.reports_count !== reports.length) {
    throw new Error(`${branchLabel} manifest reports_count mismatch`);
  }

  reports.forEach((report) => {
    requireObject(`${branchLabel} manifest report`, report);
    requireString(report, "branch_key", { expected: branchKey });
    requireString(report, "branch_label", { expected: branchLabel });
    requireString(report, "data_path");
    requiredReportKeys.forEach((entry) => {
      if (entry.type === "string") {
        requireString(report, entry.key, { allowEmpty: entry.allowEmpty });
      } else if (entry.type === "number") {
        requireNumber(report, entry.key);
      } else if (entry.type === "array") {
        requireArray(report, entry.key);
      } else if (entry.type === "boolean") {
        requireBoolean(report, entry.key);
      }
    });
  });

  return manifest;
}

function validateBranchReport(report, requiredEntries) {
  requireObject("branch report", report);
  requireString(report, "generated_at");
  requireString(report, "source_url");
  requiredEntries.forEach((entry) => {
    if (entry.type === "string") {
      requireString(report, entry.key, { allowEmpty: entry.allowEmpty });
    } else if (entry.type === "number") {
      requireNumber(report, entry.key);
    } else if (entry.type === "array") {
      requireArray(report, entry.key);
    } else if (entry.type === "boolean") {
      requireBoolean(report, entry.key);
    }
  });
  return report;
}

export function validateDailyManifest(manifest) {
  validateBranchManifest(manifest, {
    branchKey: "daily",
    branchLabel: "Cool Daily",
    requiredReportKeys: [
      { key: "slug", type: "string" },
      { key: "report_date", type: "string" },
      { key: "category", type: "string" },
      { key: "total_papers", type: "number" },
      { key: "classifier", type: "string" },
      { key: "generated_at", type: "string" },
      { key: "source_url", type: "string" },
      { key: "focus_topics", type: "array" },
      { key: "top_topics", type: "array" },
    ],
  });
  requireArray(manifest, "category_order");
  requireArray(manifest, "latest_by_category");
  return manifest;
}

export function validateDailyReport(report) {
  return validateBranchReport(report, [
    { key: "report_date", type: "string" },
    { key: "category", type: "string" },
    { key: "classifier", type: "string" },
    { key: "total_papers", type: "number" },
    { key: "focus_topics", type: "array" },
    { key: "topic_distribution", type: "array" },
    { key: "papers", type: "array" },
  ]);
}

export function validateHfManifest(manifest) {
  return validateBranchManifest(manifest, {
    branchKey: "hf-daily",
    branchLabel: "HF Daily",
    requiredReportKeys: [
      { key: "slug", type: "string" },
      { key: "report_date", type: "string" },
      { key: "total_papers", type: "number" },
      { key: "classifier", type: "string" },
      { key: "generated_at", type: "string" },
      { key: "source_url", type: "string" },
      { key: "focus_topics", type: "array" },
      { key: "top_topics", type: "array" },
      { key: "top_submitters", type: "array" },
    ],
  });
}

export function validateHfReport(report) {
  return validateBranchReport(report, [
    { key: "report_date", type: "string" },
    { key: "classifier", type: "string" },
    { key: "total_papers", type: "number" },
    { key: "topic_distribution", type: "array" },
    { key: "top_submitters", type: "array" },
    { key: "papers", type: "array" },
  ]);
}

export function validateConferenceManifest(manifest) {
  return validateBranchManifest(manifest, {
    branchKey: "conference",
    branchLabel: "Conference",
    requiredReportKeys: [
      { key: "slug", type: "string" },
      { key: "venue", type: "string" },
      { key: "venue_series", type: "string" },
      { key: "venue_year", type: "string" },
      { key: "total_papers", type: "number" },
      { key: "classifier", type: "string" },
      { key: "generated_at", type: "string" },
      { key: "source_url", type: "string" },
      { key: "subject_distribution", type: "array" },
      { key: "top_topics", type: "array" },
    ],
  });
}

export function validateConferenceReport(report) {
  return validateBranchReport(report, [
    { key: "venue", type: "string" },
    { key: "venue_series", type: "string" },
    { key: "venue_year", type: "string" },
    { key: "classifier", type: "string" },
    { key: "total_papers", type: "number" },
    { key: "is_complete", type: "boolean" },
    { key: "subject_distribution", type: "array" },
    { key: "topic_distribution", type: "array" },
    { key: "papers", type: "array" },
  ]);
}

export function validateTrendingManifest(manifest) {
  return validateBranchManifest(manifest, {
    branchKey: "trending",
    branchLabel: "Trending",
    requiredReportKeys: [
      { key: "slug", type: "string" },
      { key: "snapshot_date", type: "string" },
      { key: "since", type: "string" },
      { key: "total_repositories", type: "number" },
      { key: "generated_at", type: "string" },
      { key: "source_url", type: "string" },
      { key: "top_languages", type: "array" },
      { key: "top_repositories", type: "array" },
    ],
  });
}

export function validateTrendingReport(report) {
  return validateBranchReport(report, [
    { key: "snapshot_date", type: "string" },
    { key: "since", type: "string" },
    { key: "total_repositories", type: "number" },
    { key: "language_distribution", type: "array" },
    { key: "top_repositories", type: "array" },
    { key: "repositories", type: "array" },
  ]);
}

export function validateMagazineManifest(manifest) {
  return validateBranchManifest(manifest, {
    branchKey: "magazine",
    branchLabel: "Magazine",
    requiredReportKeys: [
      { key: "slug", type: "string" },
      { key: "issue_number", type: "number" },
      { key: "issue_title", type: "string" },
      { key: "sync_date", type: "string" },
      { key: "sections_count", type: "number" },
      { key: "generated_at", type: "string" },
      { key: "source_url", type: "string" },
      { key: "cover_image_url", type: "string", allowEmpty: true },
      { key: "excerpt", type: "string", allowEmpty: true },
      { key: "headings", type: "array" },
    ],
  });
}

export function validateMagazineReport(report) {
  validateBranchReport(report, [
    { key: "report_kind", type: "string" },
    { key: "sync_date", type: "string" },
    { key: "issue_number", type: "number" },
    { key: "issue_slug", type: "string" },
    { key: "issue_title", type: "string" },
    { key: "raw_url", type: "string" },
    { key: "cover_image_url", type: "string", allowEmpty: true },
    { key: "excerpt", type: "string", allowEmpty: true },
    { key: "lead_markdown", type: "string", allowEmpty: true },
    { key: "sections_count", type: "number" },
    { key: "headings", type: "array" },
    { key: "sections", type: "array" },
  ]);
  requireString(report, "report_kind", { expected: "magazine" });
  return report;
}

export function validateBranchCatalogManifest(manifest) {
  requireObject("branch catalog manifest", manifest);
  requireString(manifest, "generated_at");
  requireNumber(manifest, "reports_count");
  const branches = requireArray(manifest, "branches");
  const reports = requireArray(manifest, "reports");
  if (manifest.reports_count !== reports.length) {
    throw new Error("branch catalog reports_count mismatch");
  }

  branches.forEach((branch) => {
    requireObject("branch catalog branch", branch);
    requireString(branch, "branch_key");
    requireString(branch, "branch_label");
    requireString(branch, "default_report_path", { allowEmpty: true });
    requireString(branch, "manifest_path");
    requireNumber(branch, "reports_count");
  });

  reports.forEach((report) => {
    requireObject("branch catalog report", report);
    requireString(report, "branch_key");
    requireString(report, "branch_label");
    requireString(report, "data_path");
    requireString(report, "search_text");
  });

  return manifest;
}
